src/table/boot.rs

```diff
@@ -307,31 +315,72 @@ impl BootServices {
         &self,
         event_ty: EventType,
         notify_tpl: Tpl,
-        notify_fn: Option<fn(Event)>,
+        notify_fn: Option<
+            unsafe extern "efiapi" fn(event: Event, context: Option<NonNull<c_void>>),
+        >,
```

Comment:
In the original function, I couldn't see a way to pass a notification context pointer, as noted in issue #292 (I probably should've copied the text here after I closed the issue). After modifying the function to take a notification context and pass it to the notification function, I didn't see the need for the trampoline any longer. 

This might come down to policy for `uefi-rs`? Should we make it as convenient as possible for callers by letting them pass a regular rust fn and then trampoline it to an `extern "efiapi" fn()` ourselves? I can see the arguments for it, if that's what the policy is I'll need to go in and change some things in the `DebugSupport` protocol, but there's no issue there.

You're right that I can clean up the signature by using `EventNotifyFn`, there's a few places I need to do that still.

Text of issue #292:

> Hey, I'm not quite able to wrap my head around the current implementation of BootServices::create_event(). How would we properly pass context into the notification function, such as in this example?
> 
> Based on that example, the notification context is intended to be an opaque pointer to whatever the function requires, which the notification function will cast back to the concrete type. In this case, the SimpleTextInputEx protocol. I currently don't see a way to pass in any context when creating an event.

Example from EDK2 docs that passes context: https://edk2-docs.gitbook.io/edk-ii-uefi-driver-writer-s-guide/5_uefi_services/51_services_that_uefi_drivers_commonly_use/515_event_services#example-47-create-and-close-a-wait-event

Reply:
I was also considering taking a `Box<Protocol>` and converting that to a `Some(NonNull<c_void>)` instead of making the caller do it, but I had two reservations:

1. The UEFI spec isn't exactly specific on what `notify_ctx` can be. From their examples and source code, it appears to always be a `Protocol`, but it's not explicit in the docs.
2. Is it safe to convert a trait object to a single pointer in Rust? I've seen a few examples involving `Box<Box<trait T>>` and `Box::from_raw()`, but I feel that I need to dive into it a bit more to be sure.